File: src/scrapers/parisMusees.js
"use strict";

// Scraper générique de la plateforme « Paris Musées » (Drupal), deux générations
// de thème : ancien (.showcase) et récent (.card). Port de scripts/scrapers/paris_musees.py.

const base = require("./base");
const { todayIso } = require("../lib/dates");

const ACTIVITY_CATS = new Set([
  "visite", "visites", "atelier", "ateliers", "evenement", "evenements",
  "événement", "événements", "promenade", "spectacle", "conference", "conférence",
]);

function isActivity(tags) {
  return tags.some((t) => ACTIVITY_CATS.has(t.trim().toLowerCase()));
}

function iso(value) {
  if (!value) return null;
  const s = String(value).slice(0, 10);
  return /^\d{4}-\d{2}-\d{2}$/.test(s) ? s : null;
}

async function description(url) {
  try {
    return base.longestParagraph(await base.fetchText(url));
  } catch (e) {
    return null;
  }
}

// Ancien thème : .showcase
function parseShowcase($, card) {
  const c = $(card);
  let a = c.find("p.title a").first();
  if (!a.length) a = c.find("a[href]").first();
  if (!a.length) return null;
  const img = c.find("figure img").first();
  const sub = c.find("p.author").first();
  return {
    href: a.attr("href") || "",
    title: a.text().replace(/\s+/g, " ").trim(),
    subtitle: sub.length ? sub.text().replace(/\s+/g, " ").trim() : null,
    image: img.length ? img.attr("src") : null,
    date_start: iso(c.find("span.date-display-start").attr("content")),
    date_end: iso(c.find("span.date-display-end").attr("content")),
    tags: c.find("p.category").map((i, el) => $(el).text().replace(/\s+/g, " ").trim()).get(),
  };
}

// Thème récent : .card / a.card-link
function parseCard($, card) {
  const c = $(card);
  let a = c.find("a.card-link").first();
  if (!a.length) a = c.find("a[href]").first();
  const title = c.find(".card-title").first();
  if (!a.length || !title.length) return null;
  const img = c.find(".card-figure img").first();
  const sub = c.find(".card-subtitle").first();
  const cat = c.find(".meta-category").first();
  const times = c.find("time[datetime]");
  return {
    href: a.attr("href") || "",
    title: title.text().replace(/\s+/g, " ").trim(),
    subtitle: sub.length ? sub.text().replace(/\s+/g, " ").trim() : null,
    image: img.length ? img.attr("src") : null,
    date_start: times.length ? iso($(times[0]).attr("datetime")) : null,
    date_end: times.length > 1 ? iso($(times[times.length - 1]).attr("datetime")) : null,
    tags: cat.length ? [cat.text().replace(/\s+/g, " ").trim()] : [],
  };
}

async function scrape(museum) {
  const site = base.siteBase(museum.expos_url);
  const $ = base.soup(await base.fetchText(museum.expos_url));
  const today = todayIso();

  let cards = $(".showcase").toArray();
  let parse = parseShowcase;
  if (!cards.length) {
    cards = $(".card").toArray();
    parse = parseCard;
  }

  const items = [];
  const seen = new Set();
  for (const card of cards) {
    const raw = parse($, card);
    if (!raw || !raw.href || !raw.title) continue;
    if (!raw.date_start || isActivity(raw.tags)) continue;
    if (raw.date_end && raw.date_end < today) continue; // passée
    if (seen.has(raw.href)) continue;
    seen.add(raw.href);

    const url = base.absolute(site, raw.href);
    items.push({
      title: raw.title,
      subtitle: raw.subtitle,
      description: (await description(url)) || raw.subtitle,
      url,
      image_url: base.absolute(site, raw.image),
      date_start: raw.date_start,
      date_end: raw.date_end,
      tags: raw.tags.filter(Boolean),
      ext_key: raw.href.replace(/\/+$/, "").split("/").pop(),
    });
  }
  return items;
}

module.exports = { scrape };

File: scripts/scrapers/paris_musees.py
"""Scraper générique de la plateforme « Paris Musées » (Drupal).

Couvre les musées de la Ville de Paris : Carnavalet, Petit Palais, Musée d'Art
Moderne, Cognacq-Jay, Zadkine, Bourdelle, Cernuschi, Galliera, Maison de Victor
Hugo, Maison de Balzac, Vie romantique…

La plateforme existe en DEUX générations de thème ; on gère les deux :
  - ancien : cartes `.showcase` + `span.date-display-start[content]`
  - récent : cartes `.card` (a.card-link) + `<time datetime="…">`

Le `expos_url` de chaque musée (en base, éditable dans l'admin) pointe vers SA
page d'expositions. Les expositions passées (date de fin dépassée) sont ignorées.
"""
from datetime import date
from urllib.parse import urlsplit

from . import base


# Catégories qui ne sont PAS des expositions (agenda d'activités).
_ACTIVITY_CATS = {
    "visite", "visites", "atelier", "ateliers", "evenement", "evenements",
    "événement", "événements", "promenade", "spectacle", "conference", "conférence",
}


def _site_base(url):
    p = urlsplit(url)
    return f"{p.scheme}://{p.netloc}"


def _is_activity(tags):
    return any(t.strip().lower() in _ACTIVITY_CATS for t in tags)


def _iso(value):
    if not value:
        return None
    try:
        return date.fromisoformat(value[:10])
    except ValueError:
        return None


def _description(url):
    try:
        return base.longest_paragraph(base.fetch(url))
    except Exception:
        return None


def _parse_showcase(card):
    """Ancien thème : .showcase."""
    a = card.select_one("p.title a") or card.select_one("a[href]")
    if not a:
        return None
    img = card.select_one("figure img")
    sub = card.select_one("p.author")
    return {
        "href": a.get("href", ""),
        "title": a.get_text(" ", strip=True),
        "subtitle": sub.get_text(" ", strip=True) if sub else None,
        "image": img.get("src") if img else None,
        "date_start": _iso((card.select_one("span.date-display-start") or {}).get("content")),
        "date_end": _iso((card.select_one("span.date-display-end") or {}).get("content")),
        "tags": [c.get_text(" ", strip=True) for c in card.select("p.category")],
    }


def _parse_card(card):
    """Thème récent : .card / a.card-link."""
    a = card.select_one("a.card-link") or card.select_one("a[href]")
    title = card.select_one(".card-title")
    if not a or not title:
        return None
    img = card.select_one(".card-figure img")
    sub = card.select_one(".card-subtitle")
    cat = card.select_one(".meta-category")
    times = card.select("time[datetime]")
    return {
        "href": a.get("href", ""),
        "title": title.get_text(" ", strip=True),
        "subtitle": sub.get_text(" ", strip=True) if sub else None,
        "image": img.get("src") if img else None,
        "date_start": _iso(times[0]["datetime"]) if times else None,
        "date_end": _iso(times[-1]["datetime"]) if len(times) > 1 else None,
        "tags": [cat.get_text(" ", strip=True)] if cat else [],
    }


def scrape(museum):
    site = _site_base(museum.expos_url)
    page = base.soup(base.fetch(museum.expos_url))
    today = date.today()

    cards = page.select(".showcase")
    parse = _parse_showcase
    if not cards:
        cards = page.select(".card")
        parse = _parse_card

    items, seen = [], set()
    for card in cards:
        raw = parse(card)
        if not raw or not raw["href"] or not raw["title"]:
            continue
        # Une exposition a une date de début ; on écarte nav et activités d'agenda.
        if not raw["date_start"] or _is_activity(raw["tags"]):
            continue
        if raw["date_end"] and raw["date_end"] < today:   # passée -> on ignore
            continue
        if raw["href"] in seen:                            # cartes dupliquées
            continue
        seen.add(raw["href"])

        url = base.absolute(site, raw["href"])
        items.append({
            "title": raw["title"],
            "subtitle": raw["subtitle"],
            "description": _description(url) or raw["subtitle"],
            "url": url,
            "image_url": base.absolute(site, raw["image"]),
            "date_start": raw["date_start"],
            "date_end": raw["date_end"],
            "tags": [t for t in raw["tags"] if t],
            "ext_key": raw["href"].rstrip("/").rsplit("/", 1)[-1],
        })
    return items
